### ISSUE TITLE
vulkan demo hangs when resizing

### ISSUE THREAD

**Issue by christianrauch** [libsdl-org/SDL#4595]
The `testvulkan` program hangs when the window is resized (X11, wayland).

1. `./testvulkan --resizable`
2. quickly resize the window, either interactively with the mouse, or via Ctrl+Shift+{-, +}

Stack trace:
```C
#0  0x00007ffff7c8c50b in ioctl () at ../sysdeps/unix/syscall-template.S:78
#1  0x00007fffec16fc20 in intel_ioctl (arg=0x7fffffffda30, request=3223872714, fd=14) at ../src/intel/common/intel_gem.h:70
#2  anv_gem_syncobj_timeline_wait (device=device@entry=0x555555e8f070, handles=handles@entry=0x7fff840010e8, points=points@entry=0x7fff840010e0, num_items=num_items@entry=1, abs_timeout_ns=<optimized out>, wait_all=wait_all@entry=true, wait_materialize=true) at ../src/intel/vulkan/anv_gem.c:713
#3  0x00007fffec1bebe9 in anv_QueuePresentKHR (_queue=0x555555aa89a0, pPresentInfo=0x7fffffffdb20) at ../src/intel/vulkan/anv_wsi.c:335
#4  0x0000555555578a32 in render () at [...]/SDL/test/testvulkan.c:1098
#5  0x0000555555578cfd in main (argc=2, argv=0x7fffffffdd68) at [...]/SDL/test/testvulkan.c:1170
```

**Comment by flibitijibibo** [comment]
Looks like this is in ANV - might be something to report to Mesa?

**Comment by llandwerlin-intel** [comment]
There are number of validation layer errors showing up when resizing the window : 
```
../src/vulkan/wsi/wsi_common_x11.c:1108: Swapchain status changed to VK_SUBOPTIMAL_KHR
VUID-vkDestroyFence-fence-01120(ERROR / SPEC): msgNum: 1562993224 - Validation Error: [ VUID-vkDestroyFence-fence-01120 ] Object 0: handle = 0xfd23a600000000dc, type = VK_OBJECT_TYPE_FENCE; | MessageID = 0x5d296248 | VkFence 0xfd23a600000000dc[] is in use. The Vulkan spec states: All queue submission commands that refer to fence must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VUID-vkDestroyFence-fence-01120)
    Objects: 1
        [0] 0xfd23a600000000dc, type: 7, name: NULL
VUID-vkDestroyFence-fence-01120(ERROR / SPEC): msgNum: 1562993224 - Validation Error: [ VUID-vkDestroyFence-fence-01120 ] Object 0: handle = 0x55afe700000000dd, type = VK_OBJECT_TYPE_FENCE; | MessageID = 0x5d296248 | VkFence 0x55afe700000000dd[] is in use. The Vulkan spec states: All queue submission commands that refer to fence must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VUID-vkDestroyFence-fence-01120)
    Objects: 1
        [0] 0x55afe700000000dd, type: 7, name: NULL
VUID-vkFreeCommandBuffers-pCommandBuffers-00047(ERROR / SPEC): msgNum: 448332540 - Validation Error: [ VUID-vkFreeCommandBuffers-pCommandBuffers-00047 ] Object 0: handle = 0x5651450be5b0, type = VK_OBJECT_TYPE_COMMAND_BUFFER; | MessageID = 0x1ab902fc | Attempt to free VkCommandBuffer 0x5651450be5b0[] which is in use. The Vulkan spec states: All elements of pCommandBuffers must not be in the pending state (https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VUID-vkFreeCommandBuffers-pCommandBuffers-00047)
    Objects: 1
        [0] 0x5651450be5b0, type: 6, name: NULL
VUID-vkFreeCommandBuffers-pCommandBuffers-00047(ERROR / SPEC): msgNum: 448332540 - Validation Error: [ VUID-vkFreeCommandBuffers-pCommandBuffers-00047 ] Object 0: handle = 0x5651450bfeb0, type = VK_OBJECT_TYPE_COMMAND_BUFFER; | MessageID = 0x1ab902fc | Attempt to free VkCommandBuffer 0x5651450bfeb0[] which is in use. The Vulkan spec states: All elements of pCommandBuffers must not be in the pending state (https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VUID-vkFreeCommandBuffers-pCommandBuffers-00047)
```

I think the hang is triggered by the incorrect behavior of the application.

**Comment by christianrauch** [comment]
@llandwerlin-intel How do I use the Vulkan validation layer to check for errors? And are these errors on the client (using the API in the wrong way) or/and on the driver (implementing a Vulkan function the wrong way).

**Comment by llandwerlin-intel** [comment]
I usually export those env variables : `export VK_INSTANCE_LAYERS=VK_LAYER_KHRONOS_validation VK_DEVICE_LAYERS=VK_LAYER_KHRONOS_validation`
On Debian/Ubuntu you'll need to install the following packages : `vulkan-validationlayers`

The errors reported are coming from the application using the Vulkan API.

**Comment by christianrauch** [comment]
I just got a `MESA-INTEL: error: ../src/intel/vulkan/anv_batch_chain.c:2059: execbuf2 failed: Invalid argument (VK_ERROR_DEVICE_LOST)` while trying to reproduce this.

**Comment by llandwerlin-intel** [comment]
Can you run with `VK_LOADER_DEBUG=all`? It seems the validation layers are not loading.

**Comment by christianrauch** [comment]
@llandwerlin-intel The validation layers are loading and i get the same outputs. I just wanted to note that at some point during my test runs, I got this error message. But I could not reproduce it since then.

**Comment by llandwerlin-intel** [comment]
Ah okay, yeah it's not surprising to get a GPU hang if destroy the command buffers while the GPU is reading them.

**Comment by sulix** [comment]
I tried to look into this issue briefly today as well, but alas don't have the time to look too much further at the moment.

I have had sporadically the same device lost message. It usually appears before any validation layer messages, though:
```
MESA-INTEL: error: ../src/intel/vulkan/anv_batch_chain.c:2059: execbuf2 failed: Invalid argument (VK_ERROR_DEVICE_LOST)
```

Also, a few notes:
- I'm seeing more issues on X11 than Wayland. Both sometimes hang, but X11 seems to do so sooner or more frequently.
- Under X11 (but not Wayland/KWin), I get a number of issues with the swapchain being the wrong size:
```
VUID-VkSwapchainCreateInfoKHR-imageExtent-01274(ERROR / SPEC): msgNum: 2094043421 - Validation Error: [ VUID-VkSwapchainCreateInfoKHR-imageExtent-01274 ] Object 0: handle = 0xcff5f0, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0x7cd0911d | vkCreateSwapchainKHR() called with imageExtent = (593,379), which is outside the bounds returned by vkGetPhysicalDeviceSurfaceCapabilitiesKHR(): currentExtent = (568,351), minImageExtent = (568,351), maxImageExtent = (568,351). The Vulkan spec states: imageExtent must be between minImageExtent and maxImageExtent, inclusive, where minImageExtent and maxImageExtent are members of the VkSurfaceCapabilitiesKHR structure returned by vkGetPhysicalDeviceSurfaceCapabilitiesKHR for the surface (https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VUID-VkSwapchainCreateInfoKHR-imageExtent-01274)
```
- I can get rid of most of the ``vkDestroyFence()`` messages by adding a strategic call to ``vkWaitForFences()`` beforehand:
```
vkWaitForFences(vulkanContext.device, 1, &vulkanContext.fences[i], VK_TRUE, UINT64_MAX);
```
- Similarly, by also sprinkling a few ``vkQueueWaitIdle()`` and ``vkDeviceWaitIdle()`` calls around the place, I can get rid of all of the validation layer warnings (most of the time), but the hang on resize still eventually occurs.
- Sometimes ``testvulkan`` hangs on shutdown, too, in ``destroySwapchain()``:
```
#0  0x00007ffff7cbf9da in __futex_abstimed_wait_common64 (futex_word=futex_word@entry=0x7fffa1ffb910, expected=26424, clockid=clockid@entry=0, abstime=abstime@entry=0x0, 
    private=private@entry=128, cancel=cancel@entry=true) at ../sysdeps/nptl/futex-internal.c:74
#1  0x00007ffff7cbfa3f in __GI___futex_abstimed_wait_cancelable64 (futex_word=futex_word@entry=0x7fffa1ffb910, expected=<optimized out>, clockid=clockid@entry=0, abstime=abstime@entry=0x0, 
    private=private@entry=128) at ../sysdeps/nptl/futex-internal.c:123
#2  0x00007ffff7cb4753 in __pthread_clockjoin_ex (threadid=140735911278144, thread_return=0x0, clockid=0, abstime=0x0, block=<optimized out>) at pthread_join_common.c:102
#3  0x00007fffe7125934 in x11_swapchain_destroy (anv_chain=0xc28e90, pAllocator=0xd01690) at ../src/vulkan/wsi/wsi_common_x11.c:1495
#4  0x00007fffe56520f2 in DispatchDestroySwapchainKHR (pAllocator=<optimized out>, swapchain=<optimized out>, device=<optimized out>)
    at /usr/src/debug/vulkan-validationlayers-1.2.182.0-1.1.x86_64/layers/generated/layer_chassis_dispatch.cpp:439
#5  vulkan_layer_chassis::DestroySwapchainKHR (device=<optimized out>, swapchain=<optimized out>, pAllocator=<optimized out>)
    at /usr/src/debug/vulkan-validationlayers-1.2.182.0-1.1.x86_64/layers/generated/chassis.cpp:4809
#6  0x0000000000405377 in destroySwapchain () at testvulkan.c:762
#7  destroySwapchainAndSwapchainSpecificStuff (doDestroySwapchain=SDL_TRUE) at testvulkan.c:976
#8  shutdownVulkan () at testvulkan.c:1012
#9  quit (rc=rc@entry=0) at testvulkan.c:213
#10 0x000000000040443d in main (argc=<optimized out>, argv=0x7fffffffdae0) at testvulkan.c:1184
```

tl;dr: There are definitely some problems with ``testvulkan`` with things being destroyed while still in use, and with the swapchain not being resized at quite the right point relative to the window resize. Maybe there's also a driver and/or Validation Layer issue on top of these, as I definitely can get a hang without any validation layer warnings showing up first.

**Comment by llandwerlin-intel** [comment]
I feel your pain. Swapchains are indeed an issue in Vulkan. It's hard to tell when you can actually destroy them.

I might give it a try to see if I can at least avoid the hangs.

**Comment by slouken** [comment]
SDL 2.0 is now in maintenance mode, and all inactive issues are being closed. If this issue is impacting you, please feel free to reopen it with additional information.
